### DISCUSSION TITLE
How to start Java LSP by filetype and not when opening Neovim.

### DISCUSSION THREAD

**Discussion by merikan** [LazyVim/LazyVim#3449] [Q&A] [answered]
I have been using my own Java config (nvim-jdtls) for some years but decided to try Lazyvims Java Extra. The Java extra works great but the LSP starts immediately when I open neovim and not as expected, when opening a file with the filetype java. Any ideas how to make the LSP only trigger by file type? Or is it just me having this problem?

**Comment by dpetka2001** [comment]
I don't do Java, but I just installed the Extra and opened a Lua file and I can see in Lazy UI that `nvim-jdtls` is not loaded and that it has a `java` filetype handler.

**Comment by merikan** [comment]
Thx @dpetka2001 
Yes, it is only in Java projects this happens. It's a Maven project so I have `pom.xml` file in the root folder.
If I open my dotfiles folder the jdtls does not "load".
I realized that the lsp client (jdtls) is active in both cases but not attached to the buffer (Detected filetype: dashboard).
What actually happens is that when I open Neovim in a Java project the LspServer is building the project (which I don't want)

**Reply by dpetka2001** [reply]
Well I'm not familiar with how `nvim-jdtls` works, but you should read its docs to see if there's an option about not building the project at startup. Maybe then, you can disable this option in LazyVim Extra.

**Reply by rogeriofrsouza** [reply]
Do you have any other Java related config together with the Java Extra?
Just tested here launching Neovim in a Maven project and it didn't start `nvim-jdtls` right away.
I had this same issue using `sonarlint.nvim`, could you share some of your config?

**Reply by merikan** [reply]
@rogeriofrsouza I used the wrong branch so clearing the cache didn't help, I used my old working configuration so the text below is not correct. 🙈 
I have pushed a new branch `java-extras` that you are more than welcome to check and see if you find anything wrong. https://github.com/merikan/.dotfiles/tree/java-extras/config/nvim

No matter what folder I'm in when opening Neovim, the jdtls LSP starts building but it does not attach to the buffer.

```
 Language client log: /Users/peter/.local/state/nvim/lsp.log
 Detected filetype:   dashboard
 
 1 client(s) attached to this buffer: 
 
 Client: diagnosticls (id: 2, bufnr: [1])
 	filetypes:       
 	autostart:       true
 	root directory:  Running in single file mode.
 	cmd:             /Users/peter/.local/share/nvim/mason/bin/diagnostic-languageserver --stdio
 
 1 active client(s) not attached to this buffer: 
 
 Client: jdtls (id: 1, bufnr: [])
 	filetypes:       
 	autostart:       false
 	root directory:  /Users/peter/tmp
 	cmd:             /Users/peter/.local/share/nvim/mason/bin/jdtls

```

> 
> Thanks for testing. The only thing that is Java-related in my config is Java Extra. 
> Your question then got me thinking about whether something was in the cache, I cleared the cache ($XDG_CACHE_HOME/nvim) and now it works as intended. No LSPs are started when I start Neovim and LSPs are only started when I open the correct file type.
> Now when I open Newovim in a Java project the jdtls and sonarlint-nvim client are only started when opening a Java file, as expected.
> 
> Problem solved 🥳 Thanks for the help my Lazy friends 🙂

**Reply by merikan** [reply]
I copied [`java.lua`](https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/extras/lang/java.lua) from Lazyvim to my own configuration and then removed the last row `attach_jdtls()` and then it seems to work as I want it to. Maybe not so strange since `attach_jdtls()` actually does a `require("jdtls").start_or_attach`.
Although I have to run this config today and verify. There is a comment about a race condition that might happen.

One thing I'm not sure about is how lazy.nvim load plugins if you set the file type. It seems that it runs the config function and it is probably intended that way ;-) Although the documentation says "config is executed when the plugin loads". I tried with lazy=true but it made no difference so something seems to be triggering the plugin to load.

**Reply by dpetka2001** [reply]
From my limited knowledge with `lazy.nvim`, I believe that even if you set `lazy = true` in this case it won't matter, because `nvim-jdtls` is configured with `ft = "java"`, so the FileType handler will load the plugin when appropriate. In general when there are Lazy Event handlers present in a plugin spec (`ft`, `cmd`, `keys`, `event`) the plugin will be loaded when one of these handlers gets triggered first. 

I hope I'm not mistaken about this, but the maintainer is the most appropriate person to ask about this if you require more in-depth info.

**Reply by rogeriofrsouza** [reply, marked as answer]
Hey, just checked your config.
I had a similar issue using `sonarlint.nvim`, running a config very similar to yours.

I solved the problem setting lazy to true and a ft table, like this:
```lua
lazy = true,
ft = { "java" },
```

Here's [the issue](https://gitlab.com/schrieveslaach/sonarlint.nvim/-/issues/17) I created on sonarlint repo with the same `LspInfo` 😆

**Reply by merikan** [reply]
@rogeriofrsouza this awesome, it was sonarlint that somehow interfered. You are my hero 😁
